google/cloud/security/common/gcp_type/firewall_rule.py

```diff
@@ -66,3 +69,325 @@ def priority(self):
+
+        Returns:
+          bool: Whether ports_1 are a subset of ports_2 or not.
+        """
+        if any([a in ports_2 for a in ALL_REPRESENTATIONS]):
+            return True
+        return set(ports_1).issubset(ports_2)
+
+    @staticmethod
+    def ports_are_equal(ports_1, ports_2):
+        """Returns whether two port lists are the same.
+
+        Args:
+          ports_1 (list): A list of string port numbers.
+          ports_2 (list): A list of string port numbers.
+
+        Returns:
+          bool: Whether ports_1 have the same ports as ports_2.
+        """
+        if (any([a in ports_1 for a in ALL_REPRESENTATIONS]) and
+                any([a in ports_2 for a in ALL_REPRESENTATIONS])):
+            return True
+        return set(ports_1) == set(ports_2)
+
+    def is_equivalent(self, other):
+        """Returns whether this action and another are functionally equivalent.
+
+        Args:
+          other (FirewallAction): Another FirewallAction.
+
+        Returns:
+          bool: Whether these two FirewallActions are functionally equivalent.
+        """
+        return (self.action == other.action and
+                self.rules.keys() == other.rules.keys() and
+                all([
+                    self.ports_are_equal(
+                        self.rules.get(protocol, []),
+                        other.rules.get(protocol, []))
+                    for protocol in self.rules
+                ]))
+
+    def __lt__(self, other):
+        """Less than
+
+        Args:
+          other (FirewallAction): The FirewallAction to compare to.
+
+        Returns:
+          bool: Whether this action is a subset of the other action.
+        """
+        return (self.action == other.action and
+                (other.applies_to_all or
+                 all([
+                     self.ports_are_subset(
+                         self.expanded_rules.get(protocol, []),
+                         other.expanded_rules.get(protocol, []))
+                     for protocol in self.expanded_rules])))
+
+    def __gt__(self, other):
+        """Greater than
+
+        Args:
+          other (FirewallAction): The FirewallAction to compare to.
+
+        Returns:
+          bool: Whether this action is a superset of the other action.
+        """
+        return (self.action == other.action and
+                (self.applies_to_all or
+                 all([
+                     self.ports_are_subset(
+                         other.expanded_rules.get(protocol, []),
+                         self.expanded_rules.get(protocol, []))
+                     for protocol in other.expanded_rules])))
+
+    def __eq__(self, other):
+        """Equals
+
+        Args:
+          other (FirewallAction): The FirewallAction to compare to.
+
+        Returns:
+          bool: If this action is the exact same as the other FirewallAction.
+        """
+        return self.action == other.action and self.rules == other.rules
```

Comment:
The dicts in self.rules and other.rules could be in different orders, and the ports could be in different orders.

What about in __init__() doing something like this:

self.rules = sort_rules(firewall_rule_allowed)
...

def sort_rules(rules):
sorted_rules = []
for rule in sorted(rules, key=lambda k: k.get('IPProtocol', '')):
if 'ports' in rule:
\# Sort ports numerically, handle ranges through sorting by start port
rule['ports'] = sorted(rule['ports'], key=lambda k: int(k) if '-' not in k else int(k.split('-')[0]))
sorted_rules.append(rule)
return sorted_rules